wradlib/io/odim.py

```diff
@@ -0,0 +1,674 @@
+            for data in self.datas:
+                self.metadata[dataset][data] = {'what': {}, 'where': {},
+                                                'how': {}}
+
+    def read(self, filename, values=True, metadata=True, get=True):
+        """ Read data and metadata from an ODIM file.
+
+        Parameters
+        ----------
+        filename : str
+            name of the ODIM file to open
+        values : bool
+            switch to read the values
+        metadata : bool
+            switch to read the metadata into a dict
+        get : bool
+            switch to get the metadata into attributes
+
+        """
+
+        self.filehandler = h5py.File(filename, 'r')
+
+        self.ndataset = self.count_odim_datasets()
+        self.ndata = self.count_odim_datas()
+        self.datasets = ['dataset{}'.format(d + 1)
+                         for d in range(self.ndataset)]
+        self.datas = ['data{}'.format(d + 1) for d in range(self.ndata)]
+
+        if metadata:
+            self.metadata = {'Conventions': 'ODIM_H5/V2_2'}
+            self.read_metadata()
+            if get:
+                self.get_metadata()
+        if values:
+            self.read_values()
+
+    def count_odim_datasets(self, keys=None):
+        """Count the number of datasets in an ODIM file.
+
+        """
+        keys = self.filehandler.keys()
+        datasets = [key for key in keys if 'dataset' in key]
+        count = len(datasets)
+        return count
+
+    def count_odim_datas(self, keys=None):
+        """Count the number of datas in an ODIM file.
+
+        """
+
+        if keys is None:
+            keys = self.filehandler['dataset1'].keys()
+        datas = [key for key in keys if 'data' in key]
+        count = len(datas)
+
+        return count
+
+    def read_values(self, quantity=None, slicing=None,
+                    decode=True, precision=None, toarray=True):
+        """Read values of a given dataset and data.
+
+        Parameters
+        ----------
+        quantity : obj:`list` of :obj:`str`
+            list of quantity to select
+        slicing : numpy slice
+            data slice to extract
+        decode : bool
+            True to decode the data using metadata
+        precision : string
+            precision for the numpy array
+        toarray : bool
+            True to try to convert to an array
+        """
+        datas = self.datas
+        if quantity is not None:
+            datas = [self.datas[i] for i in range(self.ndata)
+                     if self.quantity[i] in quantity]
+            self.quantity = quantity
+            self.ndata = len(quantity)
+
+        values = []
+        for dataset in self.datasets:
+            vald = []
+            for data in datas:
+
+                val = self.filehandler[dataset][data]['data']
+                if slicing is None:
+                    val = val[:]
+                else:
+                    val = val[slicing]
+
+                vald.append(val)
```

Comment:
The problem with this numpy array within a list is, that the different 'data' sets might have different datatypes, so squeezing them into one array will fail then.

Reply:
I have to take back that, since the the different datasets will be upcasted to the minimum type required to hold the objects. Anyway, we have to think if upcasting is really wanted here, or if it might be better to leave the arrays with their respective datatypes.